peer/record.go

```diff
@@ -0,0 +1,188 @@
+package peer
+
+import (
+	"errors"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	pb "github.com/libp2p/go-libp2p-core/peer/pb"
+	"github.com/libp2p/go-libp2p-core/record"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// The domain string used for peer records contained in a SignedEnvelope.
+const PeerRecordEnvelopeDomain = "libp2p-peer-record"
+
+// The type hint used to identify peer records in a SignedEnvelope.
+// TODO: register multicodec
+var PeerRecordEnvelopePayloadType = []byte("/libp2p/peer-record")
+
+// ErrPeerIdMismatch is returned when attempting to sign a PeerRecord using a key that
+// does not match the PeerID contained in the record.
+var ErrPeerIdMismatch = errors.New("unable to sign peer record: key does not match record.PeerID")
+
+// PeerRecord contains information that is broadly useful to share with other peers,
+// either through a direct exchange (as in the libp2p identify protocol), or through
+// a Peer Routing provider, such as a DHT.
+//
+// Currently, a PeerRecord contains the public listen addresses for a peer, but this
+// is expected to expand to include other information in the future.
+//
+// PeerRecords are intended to be signed by the peer they describe, and there are no
+// public methods for marshalling and unmarshalling unsigned PeerRecords.
+//
+// To share a PeerRecord, first call Sign to wrap the record in a SignedEnvelope
+// and sign it with the local peer's private key:
+//
+//     rec := NewPeerRecord(myPeerId, myAddrs)
+//     envelope, err := rec.Sign(myPrivateKey)
+//
+// The resulting record.SignedEnvelope can be marshalled to a []byte and shared
+// publicly. As a convenience, the MarshalSigned method will produce the
+// SignedEnvelope and marshal it to a []byte in one go:
+//
+//     rec := NewPeerRecord(myPeerId, myAddrs)
```

Comment:
I'm afraid that `NewPeerRecord` will break frequently as we add more items to the peer record. Perhaps not providing a constructor, and having the consumer populate the struct directly is more future-proof?

Reply:
Makes sense 👍